Personal Commits

SV1002 opened the pull request:
- Made eInvSlot_Pistol a Unique Item slot.
- Secondary and Sidearm Slot weapons now accept attachments.
- Added a configuration option to allow unlimited perks per rank, instead of just one.
- Soldier rank ups are now configurable in when they happen, instead of them being after every mission (By default they rank up after every mission.)
- Added Utility slots 4 and 5.
- Starting Mission/Missions are now configurable.
- GenerateUnit Now considers the Character pool's Character Template before creating the soldier, instead of defaulting to Soldier. Fallbacks to Soldier if no Character Template is given, and respects the required classes character template
- Added config line and edited CharacterIsValid to have AlwaysValidCharacterTemplates, which should make the unit valid no matter what (Don't think this is needed, but just in case.)
- eTeam_One and eTeam_Two are now enabled, allowing enemies to spawn on said teams and function.

SV1002 committed:
Added StartingMissionTypes Config line

SV1002 committed:
Added eTeam_One and Two to be enabled + Starting mission config

SV1002 committed:
Generated soldiers now obey character pool template instead of always being soldier template

SV1002 committed:
Update UILadderSquadUpgradeScreen.uc

SV1002 committed:
Update ResistanceOperationOverhaul.int

SV1002 committed:
Update XComLadderOptions.ini

Favid commented on Config/XComLadderOptions.ini:
```diff
@@ -37,7 +47,47 @@ DEFAULT_CLASS_WEIGHT=20
+
 [ResistanceOperationOverhaul.XComGameState_LadderProgress_Override]
+; What mission plays at the start of the ladder. If multiple entries exist, then one is picked at random.
++StartingMissionTypes = "SabotageCC"
```
I think by default we'd want most mission types to be available as starting missions (though there are certainly a few we'd probably want to leave commented out)

SV1002 replied:
Agree, I just had it like this for an example, but will probably uncomment some easy missions for the Ops to begin with.

Favid commented on Src/ResistanceOperationOverhaul/Classes/UITLE_LadderModeMenu_Override.uc:
```diff
@@ -1302,6 +1302,12 @@ simulated function CreateNewLadder(int LadderDifficulty, bool NarrativesOn)
 	AddPlayerState = class'XComGameState_Player'.static.CreatePlayer(TacticalStartState, eTeam_Resistance);
 	BattleDataState.PlayerTurnOrder.AddItem(AddPlayerState.GetReference());
 
+	AddPlayerState = class'XComGameState_Player'.static.CreatePlayer(TacticalStartState, eTeam_One);
```
What are eTeam_One and eTeam_Two?

SV1002 replied:
eTEam_One and eTeam_Two are additional teams (Originally MP teams) Added by the highlander that various mods (Namely the raider factions and the Hive mod) use to assign their units to to be enemies to both ADVENT and XCOM without using eTeam_TheLost. In future when I or someone else finds a way to potentially add sit-reps to each mission, It would allow raider incursions/the Hive to work properly instead of spawning on an invalid team.

Favid commented on Src/ResistanceOperationOverhaul/Classes/UILadderSquadUpgradeScreen.uc:
```diff
@@ -728,21 +743,9 @@ simulated function SetSoldierGear()
-		mc.QueueString(equippedItem.GetMyTemplate().strImage);
-		mc.QueueString(equippedItem.GetMyTemplate().GetItemFriendlyNameNoStats());
-	}
+	equippedItem = Soldier.GetItemInSlot(eInvSlot_SecondaryWeapon, NewGameState, false);
```
Did you confirm that these changes don't cause issues with Reapers and their Claymore?

SV1002 replied:
Tested again, It works fine. Displays the claymore correctly, and switches to a new weapon Icon (Which didn't before when it was changed) when switched away from a claymore. I made it this way cause with Proficiency Reapers, It would always display a claymore despite it being a pistol.

Favid commented on Src/ResistanceOperationOverhaul/Classes/UILadderSquadUpgradeScreen.uc:
```diff
@@ -226,7 +238,7 @@ simulated function InitScreen(XComPlayerController InitController, UIMovie InitM
 			}
 			
-			for (RankIndex = 1; RankIndex < LadderData.LadderRung && RankIndex < NewSoldier.GetSoldierClassTemplate().GetMaxConfiguredRank(); RankIndex++)
+			for (RankIndex = 1; RankIndex < 2; RankIndex++)
```
This 'loop' only runs once now so it sounds like it will cause issues if a soldier joins late in the Op, because it won't rank them up multiple times if needed

SV1002 replied:
Yeah, Will fix sorry.

Favid commented on Src/ResistanceOperationOverhaul/Classes/UILadderSquadUpgradeScreen.uc:
```diff
@@ -266,7 +278,13 @@ simulated function InitScreen(XComPlayerController InitController, UIMovie InitM
+			Soldier.GetSoldierRank() == 4 && LadderData.LadderRung > default.ForthPromotionLevel || 
+			Soldier.GetSoldierRank() == 5 && LadderData.LadderRung > default.FifthPromotionLevel || 
+			Soldier.GetSoldierRank() == 6 && LadderData.LadderRung > default.SixthPromotionLevel || 
+			Soldier.GetSoldierRank() == 7 && LadderData.LadderRung > default.SeventhPromotionLevel && Soldier.GetSoldierRank() != Soldier.GetSoldierClassTemplate().GetMaxConfiguredRank()))
```
Does this work properly for classes that have the Brigadier rank (like Shadow Ops and the LW2 classes?)

SV1002 replied:
Yes, Having a mix of classes that cap at Brigadier and Colonel in my runs, they cap at their respective rank fine, With the Brig rank classes leveling up at the correct level while the Colonels remain colonel as intended.

Favid commented on Src/ResistanceOperationOverhaul/Classes/UILadderSquadUpgradeScreen.uc:
```diff
@@ -1240,6 +1274,90 @@ simulated function UpdateDataSoldierOptions()
+	// Pistol Slot
+	EquippedItem = Soldier.GetItemInSlot(eInvSlot_Pistol, NewGameState, false);
+	GetListItem(Index).EnableNavigation();
+	GetListItem(Index).UpdateDataValue(m_Sidearm, GetInventoryDisplayText(EquippedItem), OnClickSidearm);
```
I don't think we want the Pistol Slot to show up if the player is not using a mod that adds it

SV1002 replied:
True, Originally I had something set up that would only display it if said mod was installed, but came into some issues with the function actually working half the time (Would display it for the first mission, then forget it existed for some reason.) Probably the way I implemented it, but I suspect a better method would be better in checking for it.

SV1002 committed:
Fixed Promotion errors

Made it so new soldiers are promoted to the same level as other soldiers + Rank up now considers the max class promotion level for every rank, instead of just brigadier.

SV1002 committed:
AlwaysValidCharacterTemplates Initial Fix

Fixes Character Templates listed as AlwaysValidCharacterTemplates from not being promoted to squaddie at the start of a Resistance Ops.

SV1002 committed:
Starting Missions + Valid Character Examples

Added more missions to the starting missions array that are relatively simple + added various Character Templates that should always be valid from the character pool. Also fixed the default promotion levels to be more accurate.
